test_scripts/oop_poc_table.py

```diff
@@ -0,0 +1,160 @@
+            data[name] = [bool(random.getrandbits(1)) for _ in range(10)]
+        elif type == str:
+            data[name] = [
+                "".join(random.choices(string.ascii_uppercase + string.digits, k=5))
+                for _ in range(10)
+            ]
+
+    with open(path, "w", newline="", encoding="utf-8") as csvfile:
+        writer = csv.writer(csvfile)
+
+        # Write column names
+        writer.writerow(data.keys())
+
+        # Write data
+        for i in range(10):
+            writer.writerow([values[i] for values in data.values()])
+
+
+@tracer.start_as_current_span("Table")
+async def store_table():
+    # Creating annotations for my table ==================================================
+    annotations_for_my_table = {
+        "my_key_string": AnnotationsValue(
+            type=AnnotationsValueType.STRING, value=["b", "a", "c"]
+        ),
+        "my_key_bool": AnnotationsValue(
+            type=AnnotationsValueType.BOOLEAN, value=[False, False, False]
+        ),
+        "my_key_double": AnnotationsValue(
+            type=AnnotationsValueType.DOUBLE, value=[1.2, 3.4, 5.6]
+        ),
+        "my_key_long": AnnotationsValue(
+            type=AnnotationsValueType.LONG, value=[1, 2, 3]
+        ),
+        "my_key_timestamp": AnnotationsValue(
+            type=AnnotationsValueType.TIMESTAMP_MS, value=[1701362964066, 1577862000000]
+        ),
+    }
+
+    # Creating columns for my table ======================================================
+    columns = [
+        Column(id=None, name="my_string_column", column_type=ColumnType.STRING),
+        Column(id=None, name="my_integer_column", column_type=ColumnType.INTEGER),
+        Column(id=None, name="my_double_column", column_type=ColumnType.DOUBLE),
+        Column(id=None, name="my_boolean_column", column_type=ColumnType.BOOLEAN),
+    ]
+
+    # Creating a table ===================================================================
+    table = Table(
+        name="my_first_test_table",
+        columns=columns,
+        parent_id=PROJECT_ID,
+        annotations=annotations_for_my_table,
+    )
+
+    table = await table.store_schema()
+
+    print(table)
+
+    # Getting a table ====================================================================
+    copy_of_table = Table(id=table.id)
+
+    copy_of_table = await copy_of_table.get()
+
+    print(copy_of_table)
+
+    # Storing data to a table ============================================================
+    name_of_csv = "my_csv_file_with_random_data"
+    path_to_csv = os.path.join(os.path.expanduser("~/temp"), f"{name_of_csv}.csv")
+    write_random_csv_with_data(path_to_csv)
+
+    csv_path = await copy_of_table.store_rows_from_csv(csv_path=path_to_csv)
+
+    print(csv_path)
+
+    # Querying for data from a table =====================================================
+    destination_csv_location = os.path.expanduser("~/temp/my_query_results")
+
+    await Table.query(
```

Comment:
Somthing about using a method from the `Table` class outside the context of an instance of that class, but then feeding an attribute of the instance `copy_of_table` rubs me the wrong way. 

Not sure how else it should be done though since you have to pass the query string as an argument to the `query` function somehow.

Reply:
Yeah this is one of the things that I was struggling to figure out the best practice for. One thing to clarify is that we are not passing an instance of `copy_of_table`, in this case I am just grabbing the ID of the table. This is equivalent to these lines:

```
    table_id_to_query = copy_of_table.id
    await Table.query(
        query=f"SELECT * FROM {table_id_to_query}",
        result_format=CsvResultFormat(download_location=destination_csv_location),
    )
```

Reply:
Right, not an instance itself, but an attribute of that instance. 

It would feel better if you could do something like
```
copy_of_table.query(query_args)
```
So that it reads like you are querying the table you have already. But formatting the query could be weird and it's not like we don't want to let people do `Table.query` to query a table already in Synapse that they don't care to read into a Python object.

Reply:
This is very much a future work/off topic thing, but what if we didn't make people pass the whole query string themselves? 

We could use something like [jinjasql](https://github.com/sripathikrishnan/jinjasql) to allow people to pass parameters that the `Table` class could format into a query. Then someone could either pass the id of a table they want in an uninstantiated case or when there is an instance of the `Table` class we are already using, we just grab the `id` on the back end and plug it into the query.

Unfortunately, it looks like `jinjasql` has not been maintained for a while, but maybe there is something else out there that could provide similar functionality.

Reply:
Interesting library. My concerns when we get to solutioning on this are around continuing to provide the same if not more functionality for folks querying for data.

I had considered something like a "reference to my class instance" string that you could add into a query like:
`select * from :my-table-instance-id`, but that relying on those querying for the data to add in that string constant.

By not tying the query to a specific instance of a Table it makes things like joins easier to work with. SQL can be very complex and it'd be hard to implement everything we want into a Python class or semi-structured way to doing it.